File: data_baseline/openrouter__qwen__qwen3.7-plus/haskell/rep5/Main.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

module Main where

import qualified Data.Map.Strict as M
import Data.IORef
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Lazy as LT
import Data.Aeson
import GHC.Generics
import Web.Scotty
import Network.Wai (pathInfo)
import Network.HTTP.Types (Status, status201, status204, status400, status401, status404, status409)
import qualified Data.UUID.V4 as UUID
import qualified Data.UUID as UUID
import Data.Time.Clock (getCurrentTime)
import Data.Time.Format (formatTime, defaultTimeLocale)
import qualified Crypto.Hash as Hash
import qualified Data.ByteString.Char8 as C8
import Control.Monad.IO.Class (liftIO)
import Data.Maybe (fromMaybe)
import Text.Read (readMaybe)
import Data.List (sortBy, isPrefixOf)
import System.Environment (getArgs)

data ServerState = ServerState
  { stNextUserId :: Int
  , stUsers :: M.Map Int User
  , stUsersByName :: M.Map Text Int
  , stNextTodoId :: Int
  , stTodos :: M.Map Int Todo
  , stSessions :: M.Map Text Int
  } deriving (Show)

data User = User
  { uId :: Int
  , uUsername :: Text
  , uPasswordHash :: Text
  } deriving (Show, Generic)

data Todo = Todo
  { tId :: Int
  , tUserId :: Int
  , tTitle :: Text
  , tDescription :: Text
  , tCompleted :: Bool
  , tCreatedAt :: Text
  , tUpdatedAt :: Text
  } deriving (Show, Generic)

initialState :: ServerState
initialState = ServerState
  { stNextUserId = 1
  , stUsers = M.empty
  , stUsersByName = M.empty
  , stNextTodoId = 1
  , stTodos = M.empty
  , stSessions = M.empty
  }

instance ToJSON User where
  toJSON u = object ["id" .= uId u, "username" .= uUsername u]

instance ToJSON Todo where
  toJSON t = object
    [ "id" .= tId t
    , "title" .= tTitle t
    , "description" .= tDescription t
    , "completed" .= tCompleted t
    , "created_at" .= tCreatedAt t
    , "updated_at" .= tUpdatedAt t
    ]

data RegisterReq = RegisterReq
  { reqUsername :: Text
  , reqPassword :: Text
  } deriving (Generic)

instance FromJSON RegisterReq where
  parseJSON = withObject "RegisterReq" $ \v -> RegisterReq
    <$> v .: "username"
    <*> v .: "password"

data LoginReq = LoginReq
  { reqLoginUsername :: Text
  , reqLoginPassword :: Text
  } deriving (Generic)

instance FromJSON LoginReq where
  parseJSON = withObject "LoginReq" $ \v -> LoginReq
    <$> v .: "username"
    <*> v .: "password"

data PasswordChangeReq = PasswordChangeReq
  { reqOldPassword :: Text
  , reqNewPassword :: Text
  } deriving (Generic)

instance FromJSON PasswordChangeReq where
  parseJSON = withObject "PasswordChangeReq" $ \v -> PasswordChangeReq
    <$> v .: "old_password"
    <*> v .: "new_password"

data CreateTodoReq = CreateTodoReq
  { reqTitle :: Maybe Text
  , reqDescription :: Maybe Text
  } deriving (Generic)

instance FromJSON CreateTodoReq where
  parseJSON = withObject "CreateTodoReq" $ \v -> CreateTodoReq
    <$> v .:? "title"
    <*> v .:? "description"

data TodoReq = TodoReq
  { reqTodoTitle :: Maybe Text
  , reqTodoDescription :: Maybe Text
  , reqTodoCompleted :: Maybe Bool
  } deriving (Generic)

instance FromJSON TodoReq where
  parseJSON = withObject "TodoReq" $ \v -> TodoReq
    <$> v .:? "title"
    <*> v .:? "description"
    <*> v .:? "completed"

hashPassword :: Text -> Text
hashPassword pwd = T.pack $ show (Hash.hash (C8.pack $ T.unpack pwd) :: Hash.Digest Hash.SHA256)

isValidUsername :: Text -> Bool
isValidUsername u =
  let len = T.length u
      validChars = T.all (\c -> (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_') u
  in len >= 3 && len <= 50 && validChars

getCurrentTimeText :: IO Text
getCurrentTimeText = do
  now <- getCurrentTime
  let fmt = "%Y-%m-%dT%H:%M:%SZ"
  return $ T.pack $ formatTime defaultTimeLocale fmt now

jsonError :: Status -> Text -> ActionM ()
jsonError status' msg = do
  status status'
  json $ object ["error" .= msg]

requireAuth :: IORef ServerState -> ActionM (Maybe Int)
requireAuth stateRef = do
  mSession <- getCookie "session_id"
  case mSession of
    Nothing -> return Nothing
    Just sid -> do
      st <- liftIO $ readIORef stateRef
      case M.lookup sid (stSessions st) of
        Nothing -> return Nothing
        Just uid -> return (Just uid)

getTodoId :: ActionM (Maybe Int)
getTodoId = do
  req <- request
  let segments = pathInfo req
  case segments of
    ["todos", tidStr] -> return $ readMaybe (T.unpack tidStr)
    _ -> return Nothing

parseArgs :: [String] -> [(String, String)]
parseArgs [] = []
parseArgs ("--port":p:rest) = ("port", p) : parseArgs rest
parseArgs (arg:rest)
  | "--port=" `isPrefixOf` arg = ("port", drop 6 arg) : parseArgs rest
  | otherwise = parseArgs rest

main :: IO ()
main = do
  args <- getArgs
  let port = case lookup "port" (parseArgs args) of
               Just p -> read p
               Nothing -> 8080
  stateRef <- newIORef initialState
  scotty port $ do
    post "/register" $ do
      req <- jsonData :: ActionM RegisterReq
      let uname = reqUsername req
          pwd = reqPassword req
      if not (isValidUsername uname)
        then jsonError status400 "Invalid username"
        else if T.length pwd < 8
          then jsonError status400 "Password too short"
          else do
            st <- liftIO $ readIORef stateRef
            if M.member uname (stUsersByName st)
              then jsonError status409 "Username already exists"
              else do
                let newId = stNextUserId st
                    newUser = User newId uname (hashPassword pwd)
                    newState = st
                      { stNextUserId = newId + 1
                      , stUsers = M.insert newId newUser (stUsers st)
                      , stUsersByName = M.insert uname newId (stUsersByName st)
                      }
                liftIO $ writeIORef stateRef newState
                status status201
                json (object ["id" .= uId newUser, "username" .= uUsername newUser])

    post "/login" $ do
      req <- jsonData :: ActionM LoginReq
      let uname = reqLoginUsername req
          pwd = reqLoginPassword req
      st <- liftIO $ readIORef stateRef
      case M.lookup uname (stUsersByName st) of
        Nothing -> jsonError status401 "Invalid credentials"
        Just uid -> do
          let user = stUsers st M.! uid
          if uPasswordHash user == hashPassword pwd
            then do
              uuid <- liftIO UUID.nextRandom
              let token = T.pack (UUID.toString uuid)
                  newState = st { stSessions = M.insert token uid (stSessions st) }
              liftIO $ writeIORef stateRef newState
              setHeader "Set-Cookie" ("session_id=" <> LT.fromStrict token <> "; Path=/; HttpOnly")
              json (object ["id" .= uId user, "username" .= uUsername user])
            else jsonError status401 "Invalid credentials"

    post "/logout" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just _uid -> do
          mSession <- getCookie "session_id"
          case mSession of
            Just sid -> do
              st <- liftIO $ readIORef stateRef
              let newState = st { stSessions = M.delete sid (stSessions st) }
              liftIO $ writeIORef stateRef newState
            Nothing -> return ()
          json (object [] :: Value)

    get "/me" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          st <- liftIO $ readIORef stateRef
          case M.lookup uid (stUsers st) of
            Just user -> json (object ["id" .= uId user, "username" .= uUsername user])
            Nothing -> jsonError status401 "Authentication required"

    put "/password" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          req <- jsonData :: ActionM PasswordChangeReq
          st <- liftIO $ readIORef stateRef
          case M.lookup uid (stUsers st) of
            Just user -> do
              if uPasswordHash user /= hashPassword (reqOldPassword req)
                then jsonError status401 "Invalid credentials"
                else if T.length (reqNewPassword req) < 8
                  then jsonError status400 "Password too short"
                  else do
                    let newUser = user { uPasswordHash = hashPassword (reqNewPassword req) }
                        newState = st { stUsers = M.insert uid newUser (stUsers st) }
                    liftIO $ writeIORef stateRef newState
                    json (object [] :: Value)
            Nothing -> jsonError status401 "Authentication required"

    get "/todos" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          st <- liftIO $ readIORef stateRef
          let userTodos = M.elems $ M.filter (\t -> tUserId t == uid) (stTodos st)
              sortedTodos = sortBy (\a b -> compare (tId a) (tId b)) userTodos
          json sortedTodos

    post "/todos" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          req <- jsonData :: ActionM CreateTodoReq
          case reqTitle req of
            Nothing -> jsonError status400 "Title is required"
            Just title -> do
              if T.null title
                then jsonError status400 "Title is required"
                else do
                  st <- liftIO $ readIORef stateRef
                  let newId = stNextTodoId st
                  timeText <- liftIO getCurrentTimeText
                  let newTodo = Todo
                        { tId = newId
                        , tUserId = uid
                        , tTitle = title
                        , tDescription = fromMaybe "" (reqDescription req)
                        , tCompleted = False
                        , tCreatedAt = timeText
                        , tUpdatedAt = timeText
                        }
                      newState = st
                        { stNextTodoId = newId + 1
                        , stTodos = M.insert newId newTodo (stTodos st)
                        }
                  liftIO $ writeIORef stateRef newState
                  status status201
                  json newTodo

    get "/todos/:id" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          mTid <- getTodoId
          case mTid of
            Nothing -> jsonError status404 "Todo not found"
            Just tid -> do
              st <- liftIO $ readIORef stateRef
              case M.lookup tid (stTodos st) of
                Just todo | tUserId todo == uid -> json todo
                _ -> jsonError status404 "Todo not found"

    put "/todos/:id" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          mTid <- getTodoId
          case mTid of
            Nothing -> jsonError status404 "Todo not found"
            Just tid -> do
              st <- liftIO $ readIORef stateRef
              case M.lookup tid (stTodos st) of
                Just todo | tUserId todo == uid -> do
                  mReq <- jsonData :: ActionM (Maybe TodoReq)
                  let req = fromMaybe (TodoReq Nothing Nothing Nothing) mReq
                  case reqTodoTitle req of
                    Just t | T.null t -> jsonError status400 "Title is required"
                    _ -> do
                      timeText <- liftIO getCurrentTimeText
                      let newTodo = todo
                            { tTitle = fromMaybe (tTitle todo) (reqTodoTitle req)
                            , tDescription = fromMaybe (tDescription todo) (reqTodoDescription req)
                            , tCompleted = fromMaybe (tCompleted todo) (reqTodoCompleted req)
                            , tUpdatedAt = timeText
                            }
                          newState = st { stTodos = M.insert tid newTodo (stTodos st) }
                      liftIO $ writeIORef stateRef newState
                      json newTodo
                _ -> jsonError status404 "Todo not found"

    delete "/todos/:id" $ do
      mUid <- requireAuth stateRef
      case mUid of
        Nothing -> jsonError status401 "Authentication required"
        Just uid -> do
          mTid <- getTodoId
          case mTid of
            Nothing -> jsonError status404 "Todo not found"
            Just tid -> do
              st <- liftIO $ readIORef stateRef
              case M.lookup tid (stTodos st) of
                Just todo | tUserId todo == uid -> do
                  let newState = st { stTodos = M.delete tid (stTodos st) }
                  liftIO $ writeIORef stateRef newState
                  status status204
                _ -> jsonError status404 "Todo not found"

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep5/src/main.rs
use axum::{
    extract::{Path, State},
    http::{header, HeaderMap, HeaderValue, StatusCode},
    response::{IntoResponse, Response},
    routing::{get, post, put, delete},
    Json, Router,
};
use chrono::Utc;
use regex::Regex;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use uuid::Uuid;

#[derive(Clone)]
struct User {
    id: u64,
    username: String,
    password_hash: String,
}

#[derive(Clone)]
struct Todo {
    id: u64,
    user_id: u64,
    title: String,
    description: String,
    completed: bool,
    created_at: String,
    updated_at: String,
}

#[derive(Clone)]
struct AppState {
    users: Arc<Mutex<HashMap<u64, User>>>,
    user_id_counter: Arc<Mutex<u64>>,
    todos: Arc<Mutex<HashMap<u64, Todo>>>,
    todo_id_counter: Arc<Mutex<u64>>,
    sessions: Arc<Mutex<HashMap<String, u64>>>,
    username_to_id: Arc<Mutex<HashMap<String, u64>>>,
}

impl AppState {
    fn new() -> Self {
        Self {
            users: Arc::new(Mutex::new(HashMap::new())),
            user_id_counter: Arc::new(Mutex::new(0)),
            todos: Arc::new(Mutex::new(HashMap::new())),
            todo_id_counter: Arc::new(Mutex::new(0)),
            sessions: Arc::new(Mutex::new(HashMap::new())),
            username_to_id: Arc::new(Mutex::new(HashMap::new())),
        }
    }
}

#[derive(Deserialize)]
struct RegisterRequest {
    username: String,
    password: String,
}

#[derive(Deserialize)]
struct LoginRequest {
    username: String,
    password: String,
}

#[derive(Deserialize)]
struct PasswordChangeRequest {
    old_password: String,
    new_password: String,
}

#[derive(Deserialize)]
struct CreateTodoRequest {
    title: Option<String>,
    description: Option<String>,
}

#[derive(Deserialize)]
struct UpdateTodoRequest {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

#[derive(Serialize, Clone)]
struct UserResponse {
    id: u64,
    username: String,
}

#[derive(Serialize, Clone)]
struct TodoResponse {
    id: u64,
    title: String,
    description: String,
    completed: bool,
    created_at: String,
    updated_at: String,
}

impl From<Todo> for TodoResponse {
    fn from(todo: Todo) -> Self {
        Self {
            id: todo.id,
            title: todo.title,
            description: todo.description,
            completed: todo.completed,
            created_at: todo.created_at,
            updated_at: todo.updated_at,
        }
    }
}

#[derive(Debug)]
enum AppError {
    BadRequest(String),
    Unauthorized(String),
    Conflict(String),
    NotFound(String),
}

impl IntoResponse for AppError {
    fn into_response(self) -> Response {
        let (status, error_message) = match self {
            AppError::BadRequest(msg) => (StatusCode::BAD_REQUEST, msg),
            AppError::Unauthorized(msg) => (StatusCode::UNAUTHORIZED, msg),
            AppError::Conflict(msg) => (StatusCode::CONFLICT, msg),
            AppError::NotFound(msg) => (StatusCode::NOT_FOUND, msg),
        };
        
        let body = Json(serde_json::json!({
            "error": error_message
        }));
        
        (status, body).into_response()
    }
}

struct Auth {
    user_id: u64,
    token: String,
}

fn get_auth(headers: &HeaderMap, state: &Arc<AppState>) -> Result<Auth, AppError> {
    let cookies = headers.get(header::COOKIE);
    let session_token = if let Some(cookie_header) = cookies {
        let cookie_str = cookie_header.to_str().unwrap_or("");
        cookie_str
            .split(';')
            .find_map(|c| {
                let c = c.trim();
                if c.starts_with("session_id=") {
                    Some(c["session_id=".len()..].to_string())
                } else {
                    None
                }
            })
    } else {
        None
    };
    
    if let Some(token) = session_token {
        let sessions = state.sessions.lock().unwrap();
        if let Some(&user_id) = sessions.get(&token) {
            return Ok(Auth { user_id, token });
        }
    }
    
    Err(AppError::Unauthorized("Authentication required".to_string()))
}

fn hash_password(password: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(password.as_bytes());
    format!("{:x}", hasher.finalize())
}

async fn register(
    State(state): State<Arc<AppState>>,
    Json(req): Json<RegisterRequest>,
) -> Result<(StatusCode, Json<UserResponse>), AppError> {
    let username_re = Regex::new(r"^[a-zA-Z0-9_]{3,50}$").unwrap();
    if !username_re.is_match(&req.username) {
        return Err(AppError::BadRequest("Invalid username".to_string()));
    }
    
    if req.password.len() < 8 {
        return Err(AppError::BadRequest("Password too short".to_string()));
    }
    
    let mut username_to_id = state.username_to_id.lock().unwrap();
    if username_to_id.contains_key(&req.username) {
        return Err(AppError::Conflict("Username already exists".to_string()));
    }
    
    let mut users = state.users.lock().unwrap();
    let mut counter = state.user_id_counter.lock().unwrap();
    *counter += 1;
    let new_id = *counter;
    
    let new_user = User {
        id: new_id,
        username: req.username.clone(),
        password_hash: hash_password(&req.password),
    };
    users.insert(new_id, new_user.clone());
    username_to_id.insert(req.username.clone(), new_id);
    
    Ok((
        StatusCode::CREATED,
        Json(UserResponse {
            id: new_id,
            username: req.username,
        }),
    ))
}

async fn login(
    State(state): State<Arc<AppState>>,
    Json(req): Json<LoginRequest>,
) -> Result<(HeaderMap, Json<UserResponse>), AppError> {
    let username_to_id = state.username_to_id.lock().unwrap();
    let user_id = *username_to_id
        .get(&req.username)
        .ok_or_else(|| AppError::Unauthorized("Invalid credentials".to_string()))?;
    
    let users = state.users.lock().unwrap();
    let user = users
        .get(&user_id)
        .ok_or_else(|| AppError::Unauthorized("Invalid credentials".to_string()))?;
    
    if user.password_hash != hash_password(&req.password) {
        return Err(AppError::Unauthorized("Invalid credentials".to_string()));
    }
    
    let token = Uuid::new_v4().to_string();
    state.sessions.lock().unwrap().insert(token.clone(), user_id);
    
    let mut headers = HeaderMap::new();
    headers.insert(
        header::SET_COOKIE,
        HeaderValue::from_str(&format!("session_id={}; Path=/; HttpOnly", token)).unwrap(),
    );
    
    Ok((
        headers,
        Json(UserResponse {
            id: user.id,
            username: user.username.clone(),
        }),
    ))
}

async fn logout(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> Result<Json<serde_json::Value>, AppError> {
    let auth = get_auth(&headers, &state)?;
    state.sessions.lock().unwrap().remove(&auth.token);
    Ok(Json(serde_json::json!({})))
}

async fn get_me(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> Result<Json<UserResponse>, AppError> {
    let auth = get_auth(&headers, &state)?;
    let users = state.users.lock().unwrap();
    let user = users
        .get(&auth.user_id)
        .ok_or_else(|| AppError::Unauthorized("Authentication required".to_string()))?;
    Ok(Json(UserResponse {
        id: user.id,
        username: user.username.clone(),
    }))
}

async fn change_password(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(req): Json<PasswordChangeRequest>,
) -> Result<Json<serde_json::Value>, AppError> {
    let auth = get_auth(&headers, &state)?;
    if req.new_password.len() < 8 {
        return Err(AppError::BadRequest("Password too short".to_string()));
    }
    
    let mut users = state.users.lock().unwrap();
    let user = users
        .get_mut(&auth.user_id)
        .ok_or_else(|| AppError::Unauthorized("Authentication required".to_string()))?;
    
    if user.password_hash != hash_password(&req.old_password) {
        return Err(AppError::Unauthorized("Invalid credentials".to_string()));
    }
    
    user.password_hash = hash_password(&req.new_password);
    Ok(Json(serde_json::json!({})))
}

async fn get_todos(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> Result<Json<Vec<TodoResponse>>, AppError> {
    let auth = get_auth(&headers, &state)?;
    let todos = state.todos.lock().unwrap();
    let mut user_todos: Vec<TodoResponse> = todos
        .values()
        .filter(|t| t.user_id == auth.user_id)
        .map(|t| TodoResponse::from(t.clone()))
        .collect();
    
    user_todos.sort_by_key(|t| t.id);
    Ok(Json(user_todos))
}

async fn create_todo(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(req): Json<CreateTodoRequest>,
) -> Result<(StatusCode, Json<TodoResponse>), AppError> {
    let auth = get_auth(&headers, &state)?;
    let title = req.title.unwrap_or_default();
    if title.trim().is_empty() {
        return Err(AppError::BadRequest("Title is required".to_string()));
    }
    
    let description = req.description.unwrap_or_default();
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    
    let mut todos = state.todos.lock().unwrap();
    let mut counter = state.todo_id_counter.lock().unwrap();
    *counter += 1;
    let new_id = *counter;
    
    let new_todo = Todo {
        id: new_id,
        user_id: auth.user_id,
        title,
        description,
        completed: false,
        created_at: now.clone(),
        updated_at: now,
    };
    todos.insert(new_id, new_todo.clone());
    
    Ok((StatusCode::CREATED, Json(TodoResponse::from(new_todo))))
}

async fn get_todo(
    Path(todo_id): Path<u64>,
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> Result<Json<TodoResponse>, AppError> {
    let auth = get_auth(&headers, &state)?;
    let todos = state.todos.lock().unwrap();
    let todo = todos
        .get(&todo_id)
        .ok_or_else(|| AppError::NotFound("Todo not found".to_string()))?;
    
    if todo.user_id != auth.user_id {
        return Err(AppError::NotFound("Todo not found".to_string()));
    }
    
    Ok(Json(TodoResponse::from(todo.clone())))
}

async fn update_todo(
    Path(todo_id): Path<u64>,
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(req): Json<UpdateTodoRequest>,
) -> Result<Json<TodoResponse>, AppError> {
    let auth = get_auth(&headers, &state)?;
    let mut todos = state.todos.lock().unwrap();
    let todo = todos
        .get_mut(&todo_id)
        .ok_or_else(|| AppError::NotFound("Todo not found".to_string()))?;
    
    if todo.user_id != auth.user_id {
        return Err(AppError::NotFound("Todo not found".to_string()));
    }
    
    if let Some(ref title) = req.title {
        if title.trim().is_empty() {
            return Err(AppError::BadRequest("Title is required".to_string()));
        }
        todo.title = title.clone();
    }
    
    if let Some(description) = req.description {
        todo.description = description;
    }
    
    if let Some(completed) = req.completed {
        todo.completed = completed;
    }
    
    todo.updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    
    Ok(Json(TodoResponse::from(todo.clone())))
}

async fn delete_todo(
    Path(todo_id): Path<u64>,
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> Result<StatusCode, AppError> {
    let auth = get_auth(&headers, &state)?;
    let mut todos = state.todos.lock().unwrap();
    let todo = todos
        .get(&todo_id)
        .ok_or_else(|| AppError::NotFound("Todo not found".to_string()))?;
    
    if todo.user_id != auth.user_id {
        return Err(AppError::NotFound("Todo not found".to_string()));
    }
    
    todos.remove(&todo_id);
    Ok(StatusCode::NO_CONTENT)
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut port = 8080;
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse().expect("Invalid port number");
            i += 2;
        } else {
            i += 1;
        }
    }

    let state = Arc::new(AppState::new());

    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(get_me))
        .route("/password", put(change_password))
        .route("/todos", get(get_todos))
        .route("/todos", post(create_todo))
        .route("/todos/:id", get(get_todo))
        .route("/todos/:id", put(update_todo))
        .route("/todos/:id", delete(delete_todo))
        .with_state(state);

    let addr = format!("0.0.0.0:{}", port);
    let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
    println!("Listening on {}", addr);
    axum::serve(listener, app).await.unwrap();
}
